refreshable/src/main/scala/prometheus4cats/refreshable/InstrumentedRefreshable.scala

```diff
@@ -0,0 +1,256 @@
+package prometheus4cats.refreshable
+
+import cats.effect.kernel.{MonadCancelThrow, Resource}
+import cats.effect.kernel.syntax.monadCancel._
+import cats.effect.kernel.syntax.resource._
+import cats.syntax.flatMap._
+import cats.syntax.foldable._
+import cats.syntax.functor._
+import cats.{Applicative, Monad}
+import com.permutive.refreshable.{CachedValue, Refreshable}
+import prometheus4cats._
+import retry.RetryDetails
+
+import scala.concurrent.duration.FiniteDuration
+
+class InstrumentedRefreshable[F[_], A] private (
+    underlying: Refreshable[F, A],
+    name: String,
+    readCounter: Counter.Labelled[F, Long, (String, CachedValue[A])],
+    runningGauge: Gauge.Labelled[F, Boolean, String],
+    exhaustedRetriesGauge: Gauge.Labelled[F, Boolean, String]
+)(implicit
+    override val functor: Monad[F]
+) extends Refreshable[F, A] {
+  override def get: F[CachedValue[A]] =
+    underlying.get.flatTap(value => readCounter.inc(name -> value))
+
+  override def cancel: F[Boolean] =
+    underlying.cancel.flatTap(
+      if (_) runningGauge.set(false, name) else Applicative[F].unit
+    )
+
+  override def restart: F[Boolean] =
+    underlying.restart.flatTap(
+      if (_)
+        runningGauge.set(true, name) >> exhaustedRetriesGauge.set(false, name)
+      else Applicative[F].unit
+    )
+}
+
+object InstrumentedRefreshable {
+  private val prefix: Metric.Prefix = "refreshable"
+  private val refreshableLabelName: Label.Name = "refreshable_name"
+
+  private def metrics[F[_]: MonadCancelThrow, A](
+      name: String,
+      metricFactory: MetricFactory[F]
+  )(
+      currentOnNewValue: Option[(A, FiniteDuration) => F[Unit]],
+      currentOnRefreshFailure: PartialFunction[(Throwable, RetryDetails), F[
+        Unit
+      ]],
+      currentOnExhaustedRetries: PartialFunction[Throwable, F[Unit]]
+  ): F[
```

Comment:
Can this not return `Resource` and then we don't need to `unsafeBuild` the metrics?